src/preamp.jl

```diff
@@ -17,6 +17,7 @@ This struct is currently mutable because in the case of noise modelling based on
     gain / max_e the data was produced with and give it as simulation input.
 """
 @with_kw mutable struct GenericPreAmp <: PreAmp
+    # Note: these default T() expressions don't work cause T not defined
```

Comment:
I think we need to discuss at some point the `T` situation in general. It's all over the place now, and I remember @oschulz saying that we should make it all proper and then the user can decide `T` or have some default etc. I don't even fully get the system. We should call at some point and chat about this. So maybe for now I'd leave it like this, and then call next week?

Reply:
I have something for that now (brand new, but we'll use it in BAT a lot now):

https://oschulz.github.io/HeterogeneousComputing.jl/stable/api/#HeterogeneousComputing.GenContext

`GenContext` gives the user a way to specify desired precision and compute device (the algorithm is free to ignore them if it can only do CPU or a fixed precision) and random number generator. By passing it around internally in packages between functions, we have consistency and a single RNG seed for a whole application as well.

Reply:
Ok let's leave it for another PR, will add in Issues